streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -1477,6 +1590,67 @@ public Object getStateLock() {
         return stateLock;
     }
 
+    public Map<String, KafkaFuture<Uuid>> consumerClientInstanceIds(final Duration timeout) {
+        boolean setDeadline = false;
+
+        final Map<String, KafkaFuture<Uuid>> result = new HashMap<>();
+
+        KafkaFutureImpl<Uuid> future = new KafkaFutureImpl<>();
+        if (mainConsumerClientInstanceId != null) {
+            future.complete(mainConsumerClientInstanceId);
+        } else {
+            mainConsumerInstanceIdFuture = future;
+            setDeadline = true;
+        }
+        result.put(getName() + "-consumer", future);
+
+        future = new KafkaFutureImpl<>();
+        if (restoreConsumerClientInstanceId != null) {
+            future.complete(restoreConsumerClientInstanceId);
+        } else {
+            restoreConsumerInstanceIdFuture = future;
+            setDeadline = true;
+        }
+        result.put(getName() + "-restore-consumer", future);
+
+        if (setDeadline) {
+            fetchDeadline = time.milliseconds() + timeout.toMillis();
+        }
+
+        return result;
+    }
+
+    public KafkaFuture<Map<String, KafkaFuture<Uuid>>> producersClientInstanceIds(final Duration timeout) {
+        final KafkaFutureImpl<Map<String, KafkaFuture<Uuid>>> result = new KafkaFutureImpl<>();
+
+        if (processingMode.equals(StreamsConfigUtils.ProcessingMode.EXACTLY_ONCE_ALPHA)) {
+//            for (final TaskId taskId : taskManager.activeTaskIds()) {
+//                future = new KafkaFutureImpl<>();
+//                if (taskProducersClientInstanceIds.get(taskId) != null) {
+//                    future.complete(taskProducersClientInstanceIds.get(taskId));
+//                } else {
+//                    taskProducersInstanceIdsFuture.put(taskId, future);
+//                    setDeadline = true;
+//                }
+//                result.put(getName() + "-" + taskId + "-producer", future);
+//            };
+        } else {
+            final KafkaFutureImpl<Uuid> producerFuture = new KafkaFutureImpl<>();
+            if (threadProducerClientInstanceId != null) {
+                producerFuture.complete(threadProducerClientInstanceId);
+            } else {
+                threadProducerInstanceIdFuture = producerFuture;
+                if (fetchDeadline == -1) {
+                    fetchDeadline = time.milliseconds() + timeout.toMillis();
+                }
+            }
+
+            result.complete(Collections.singletonMap(getName() + "-producer", producerFuture));
```

Comment:
For thread-producer (ie, ALOS / EOSv2) we can complete the future we return right away because we did create all nested futures already (completion of the outer future means, all inner futures got added to the map and we haven only one inner for the thread producer case).

For EOSv1 we need to forward the request to create the nested futures (one per task) to the TaskManager, ie, we need the StreamThread to not just complete, but actually first create the nested futures (to avoid race conditions with rebalancing and task creating/removal) -- for this case the outer future will not be complete here, but only after the TM did create all task futures.

My intention is to take the created map as a snapshot, ie, if a rebalance happens in between, we won't add new futures to the map (for newly assigned tasks), and for removed tasks, we will cancel the future.

Not sure about tasks which are not active but still restoring right now.

EOSv1 is becoming messy. Any good ideas how we could simplify it?

Reply:
Could we be lazy and just treat any unavailable tasks as if they were set up with telemetry is disabled on the client itself?

Not ideal but should keep things simple. Otherwise we might timeout all the time with restoring tasks